whispers/core/utils.py

```diff
@@ -0,0 +1,216 @@
+
+    return value
+
+
+def similar_strings(a: str, b: str) -> float:
+    """Returns similarity coefficient between two strings"""
+    a = simple_string(a).replace("_", "")
+    b = simple_string(b).replace("_", "")
+
+    return ratio(a, b)
+
+
+def is_ascii(data: str) -> bool:
+    """Checks if given data is printable text"""
+    if isinstance(data, bytes):
+        try:
+            data = data.decode("utf-8")
+        except Exception:
+            return False
+
+    if not isinstance(data, (str, int)):
+        return False
+
+    for ch in str(data):
+        if ch not in string.printable:
+            return False
+
+    return True
+
+
+def is_base64(data: str) -> bool:
+    """Checks if given data is base64-decodable to text"""
+    if not isinstance(data, str):
+        return False
+
+    try:
+        b64decode(data).decode("utf-8")
+        return True
+
+    except Exception:
+        return False
+
+
+def is_base64_bytes(data: str) -> bool:
+    """Checks if given data is base64-decodable to bytes"""
+    try:
+        return b64decode(data) != b""
+
+    except Exception:
+        return False
+
+
+def is_uri(data: str) -> bool:
+    """Checks if given data resemples a URI"""
+    if not is_ascii(data):
+        return False
+
+    if isinstance(data, int):
+        return False
+
+    regex = r"(?!\s)[:\w\d]+://.+"
+
+    if any(map(lambda ch: ch in data, string.whitespace)):
+        return False
+
+    if not re.match(regex, data, flags=re.IGNORECASE):
+        return False
+
+    return True
+
+
+def is_path(data: str) -> bool:
+    """Checks if given data resemples a system path"""
+    if not is_ascii(data):
+        return False
+
+    if isinstance(data, int):
+        return False
+
+    regex = r"^((([A-Z]|file|root):)?(\.+)?[/\\]+).*$"
+
+    if not re.match(regex, data, flags=re.IGNORECASE):
+        return False
+
+    return True
+
+
+def is_iac(data: str) -> bool:
+    """Checks if given data resemples IaC function"""
+    if not is_ascii(data):
+        return False
+
+    if isinstance(data, int):
+        return False
+
+    regex = r"\![A-Za-z]+ .+"
+
+    if not re.match(regex, data, flags=re.IGNORECASE):
+        return False
+
+    return True
+
+
+def is_luhn(data: str) -> bool:
+    """Checks if given data resembles a credit card number"""
+    if not is_ascii(data):
+        return False
+
+    if not str(data).isnumeric():
+        return False
+
+    return luhn_verify(str(data))
+
+
+def is_similar(key: str, value: str, similarity: float) -> bool:
+    """
+    Checks similarity between key and value.
+    Returns True if calculated similarity is greater than given similarity.
+    """
+    return similar_strings(key, value) >= similarity
+
+
+def find_line_number(pair: KeyValuePair) -> int:
+    """Finds line number using pair keypath and value"""
+    if pair.line:
+        return pair.line  # Already set
+
+    valuepath = pair.value.split("\n")[0][:16]
```

Comment:
Why `16` here? 😄 I'm sure there's good reason, but maybe an inline comment will help

Reply:
This is for specific cases like the following https://github.com/Skyscanner/whispers/blob/9f5f6da5c54877fb2f0175bac65052c9af633fb4/tests/fixtures/privatekeys.yml#L6-L10

When this is loaded by whispers, key `key` has value of `-----BEGIN RSA KEY----- \nQyNTUxOQAAACCtrF27B/zd9DEpd38IbVBy93wSeYXKU0AGXMyO8ePu2QAAAKBSzpYEUs6W \n-----END RSA KEY-----`. However, if you were to look for that string in that file, you would not find it, because in plaintext it's multi-line is formatted differently. 

In order to find the line number where this value begins correctly, the operation is to first split the value to get `-----BEGIN RSA KEY----- `, and then only take up to the first 16 chars for finding the string in a line of plaintext.